### PR TITLE
fix: Move all componentWillUnMount functionality to respective events

### LINKED ISSUE #1515: Warning: Can't perform a React state update on an unmounted component
I am using `GraphiQL` in the form as a textarea field and every time I hit submit and navigate away from the page I get this warning

`
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
`
   `in GraphiQL (created by GraphQlTextArea)`

Can someone please look into this?

### PR DESCRIPTION
**Pull request by harshithpabbati**
Closes #1515 
Move all functionality from `componentWillUnMount` lifecycle

@acao can you review this PR

### CODE REVIEW

File: packages/graphiql/src/components/GraphiQL.tsx [hunk 1]
```diff
@@ -1133,6 +1114,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       const leftSize = moveEvent.clientX - getLeft(editorBar) - offset;
       const rightSize = editorBar.clientWidth - leftSize;
       this.setState({ editorFlex: leftSize / rightSize });
+      this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex));
```
Comment (acao) [inline packages/graphiql/src/components/GraphiQL.tsx:1117]: let's debounce this one, using `utility/debounce`.

File: packages/graphiql/src/components/GraphiQL.tsx [hunk 2]
```diff
@@ -1080,6 +1049,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
               this.docExplorerComponent.showDoc(type);
             }
           });
+          this._storage.set(
```
Comment (acao) [inline packages/graphiql/src/components/GraphiQL.tsx:1052]: let's debounce this one as well

File: packages/graphiql/src/components/GraphiQL.tsx [hunk 3]
```diff
@@ -1217,6 +1212,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
     this.setState({
       docExplorerWidth: DEFAULT_DOC_EXPLORER_WIDTH,
     });
+    this._storage.set(
```
Comment (acao) [inline packages/graphiql/src/components/GraphiQL.tsx:1215]: debounce this one

File: packages/graphiql/src/components/GraphiQL.tsx [hunk 4]
```diff
@@ -1250,6 +1249,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
           variableEditorHeight: bottomSize,
         });
       }
+      this._storage.set(
```
Comment (acao) [inline packages/graphiql/src/components/GraphiQL.tsx:1252]: debounce as well

**Comment by acao** [comment]
awesome work @harshithpabbati ! |

we just need to make sure to debounce the more high-throughput events to avoid performance/render-blocking issues, as invoking lots of storage.set() events can sometimes lead to issues like that

as for the delay, 500-1000ms should be fine. these values only need to be persisted in case the user navigates away or refreshes. this will essentially delay the write until the user has finished, or if there's lots of dragging or variables editing going on, it will only write to ls every 500ms-1000ms, rather than writing to localstorage 60 times a second or more

**Comment by harshithpabbati** [comment]
@acao should I debounce the whole function ?

**Comment by acao** [comment]
@harshithpabbati no, only the storage write events

**Comment by harshithpabbati** [comment]
> @harshithpabbati no, only the storage write events

ok, I will update that

**Comment by acao** [comment]
if you debounce the setState() events, we get weird issues with resizing panes as you can see

**Comment by harshithpabbati** [comment]
> if you debounce the setState() events, we get weird issues with resizing panes as you can see

oh, yeah true 👍

File: packages/graphiql/src/components/GraphiQL.tsx (hunk 1 above)
Comment (harshithpabbati) [reply packages/graphiql/src/components/GraphiQL.tsx:1118]: If I make it like this ` debounce(500, this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex)));`

then this error is showing up
```
Argument of type '{ isQuotaError: boolean; error: any; }' is not assignable to parameter of type '(...args: any[]) => any'.
  Type '{ isQuotaError: boolean; error: any; }' provides no match for the signature '(...args: any[]): any'.ts(2345)
```
Should I make a function for storage set and then pass the variables ?

@acao can you help me with this?

**Comment by acao** [comment]
@harshithpabbati why did you close this PR?

File: packages/graphiql/src/components/GraphiQL.tsx (hunk 1 above)
Comment (acao) [reply packages/graphiql/src/components/GraphiQL.tsx:1118]: @harshithpabbati did not see this before, sorry

**Comment by harshithpabbati** [comment]
> @harshithpabbati why did you close this PR?

That's because we found a better solution than this.

**Comment by acao** [comment]
@harshithpabbati i'm not sure that's the case, we just found a better way to reproduce the error

**Comment by harshithpabbati** [comment]
Oh is it ? Then cool!

**Comment by harshithpabbati** [comment]
> @harshithpabbati i'm not sure that's the case, we just found a better way to reproduce the error. is there a conversation that i did't see where you and Mateo decided his solution was better?

No, we didn't have any discussion

**Comment by acao** [comment]
@harshithpabbati yes, as i mentioned in discord, we will wait until both solutions are ready, and make sure we can reproduce the error with both and see if one doesn't solve it better than the other. both proposals are welcome.

**Comment by harshithpabbati** [comment]
Okay, so there is an error when I use denounce

If I make it like this debounce(500, this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex)));

then this error is showing up

Argument of type '{ isQuotaError: boolean; error: any; }' is not assignable to parameter of type '(...args: any[]) => any'.
  Type '{ isQuotaError: boolean; error: any; }' provides no match for the signature '(...args: any[]): any'.ts(2345)
Should I make a function for storage set and then pass the variables ?

@acao can you help me with this?

File: packages/graphiql/src/components/GraphiQL.tsx (hunk 1 above)
Comment (acao) [reply packages/graphiql/src/components/GraphiQL.tsx:1118]: @harshithpabbati that argument is supposed to be a function it looks like, so I think you'll want to call:

```ts
debounce(500, () => this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex)));
```

Comment (harshithpabbati) [reply packages/graphiql/src/components/GraphiQL.tsx:1118]: Okay, cool thanks @acao 

I will update the PR

**Comment by harshithpabbati** [comment]
Done updated @acao :)

File: packages/graphiql/src/components/GraphiQL.tsx [hunk 5]
```diff
@@ -1195,12 +1182,24 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
           docExplorerOpen: true,
           docExplorerWidth: Math.min(docsSize, 650),
         });
+        this._storage.set(
+          'docExplorerWidth',
+          JSON.stringify(this.state.docExplorerWidth),
```
Comment (acao) [inline packages/graphiql/src/components/GraphiQL.tsx:1187]: ah, another one that needs debounced. anything that has to do with high throughput events

File: packages/graphiql/src/components/GraphiQL.tsx [hunk 6]
```diff
@@ -1195,12 +1182,24 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
...
+        JSON.stringify(this.state.docExplorerOpen),
+      );
     };
 
     let onMouseUp: OnMouseUpFn = () => {
       if (!this.state.docExplorerOpen) {
         this.setState({ docExplorerWidth: hadWidth });
+        this._storage.set(
+          'docExplorerWidth',
+          JSON.stringify(this.state.docExplorerWidth),
```
Comment (acao) [inline packages/graphiql/src/components/GraphiQL.tsx:1201]: same here!

File: packages/graphiql/src/components/GraphiQL.tsx [hunk 7]
```diff
@@ -1033,6 +1001,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
 
   handleEditVariables = (value: string) => {
     this.setState({ variables: value });
+    this._storage.set('variables', value);
```
Comment (acao) [inline packages/graphiql/src/components/GraphiQL.tsx:1004]: also high throughput. this event is triggered on every key strike

**Comment by acao** [comment]
looking good @harshithpabbati !

as a rule of thumb, any event that has its events triggered many many times a second, such as resize or keyboard stroke, should debounce all the `localStorage.setItem()` calls (via `storage.set()`)

**Comment by harshithpabbati** [comment]
> looking good @harshithpabbati !
> 
> as a rule of thumb, any event that has its events triggered many many times a second, such as resize or keyboard stroke, should debounce all the `localStorage.setItem()` calls (via `storage.set()`)

Sure, I will update it soon

**Comment by harshithpabbati** [comment]
@acao is this PR helpful ?

**Comment by acao** [comment]
@harshithpabbati it might still be?

**Comment by acao** [comment]
lol oops! tried merging via UI, fixing it now. gonna try and see if this resolves our user's issue

**Pull request merged**